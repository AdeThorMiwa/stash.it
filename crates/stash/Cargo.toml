[package]
name = "stash"
version = "0.1.0"
edition = "2024"

[dependencies]
shared = { path = "../shared" }
serde = { workspace = true }
chrono = { workspace = true }
async-trait = { workspace = true }
more-di = { workspace = true }
thiserror = { workspace = true }
config = { workspace = true }
tokio = { workspace = true }
insta = { workspace = true }
serde_json = { workspace = true }
derive_builder = { workspace = true }
