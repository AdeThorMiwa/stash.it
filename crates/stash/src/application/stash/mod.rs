use crate::{
    application::stash::command::{CreateStashCommand, UpdateStashBalanceCommand, UpdateStashStatusCommand},
    domain::{
        events::{StashBalanceUpdatedEvent, StashCreatedEvent, StashStatusUpdatedEvent},
        repositories::StashRepository,
        stash::stash::Stash,
    },
};
use di::injectable;
use shared::infrastructure::{
    messaging::EventBus,
    types::{
        Result,
        error::{DomainError, Error},
    },
};
use std::sync::Arc;

pub mod command;

#[injectable]
pub struct StashService {
    stash_repo: Arc<dyn StashRepository>,
    event_bus: Arc<dyn EventBus>,
}

impl StashService {
    pub async fn create_stash(&self, command: CreateStashCommand) -> Result<Stash> {
        self.assert_can_create_stash(&command).await?;
        let stash = Stash::new(&command.user_id, &command.name, &command.tags);
        self.stash_repo.save(&stash).await?;
        let stash_created_event = StashCreatedEvent::new(stash.get_pid(), stash.get_user_id());
        self.event_bus.publish(stash_created_event).await?;
        Ok(stash)
    }

    pub async fn update_stash_status(&self, command: UpdateStashStatusCommand) -> Result<()> {
        let mut stash = self
            .stash_repo
            .find_by_pid(&command.stash_id)
            .await?
            .ok_or(Error::DomainError(DomainError::EntityNotFound))?;

        stash.update_status(&command.new_status);
        self.stash_repo.save(&stash).await?;
        let stash_status_updated_event = StashStatusUpdatedEvent::new(stash.get_pid(), stash.get_status());
        self.event_bus.publish(stash_status_updated_event).await?;
        Ok(())
    }

    pub async fn update_stash_balance(&self, command: UpdateStashBalanceCommand) -> Result<()> {
        let mut stash = self
            .stash_repo
            .find_by_pid(&command.stash_id)
            .await?
            .ok_or(Error::DomainError(DomainError::EntityNotFound))?;

        stash.update_balance(&command.asset, &command.new_balance);
        self.stash_repo.save(&stash).await?;
        let stash_balance_updated_event = StashBalanceUpdatedEvent::new(stash.get_pid(), &command.asset, &command.new_balance);
        self.event_bus.publish(stash_balance_updated_event).await?;
        Ok(())
    }

    async fn assert_can_create_stash(&self, command: &CreateStashCommand) -> Result<()> {
        let tag_len = command.tags.len();
        if tag_len > Self::max_tag_len() {
            return Err(Error::AssertError(format!(
                "max tags exceeded. max: {} got: {}",
                Self::max_tag_len(),
                tag_len
            )));
        }

        if self.stash_repo.exists_with_name_for_user(&command.user_id, &command.name).await? {
            return Err(Error::AssertError("stash with name already exist".to_string()));
        }

        Ok(())
    }

    /// maximum tags allowed on a single stash
    fn max_tag_len() -> usize {
        10
    }
}
