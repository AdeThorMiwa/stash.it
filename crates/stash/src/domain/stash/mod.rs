pub mod name;
pub mod stash;
pub mod status;
pub mod tag;
