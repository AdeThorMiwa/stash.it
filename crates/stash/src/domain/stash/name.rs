use std::str::FromStr;

use serde::Deserialize;
use thiserror::Error;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct StashName(String);

#[derive(Debug, Error)]
pub enum StashNameError {
    #[error("Invalid character length: {0}")]
    InvalidCharacterLength(usize),
    #[error("Stash name must start with a letter")]
    NameMustStartWithLetter,
}

impl FromStr for StashName {
    type Err = StashNameError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.len() < 3 || s.len() > 15 {
            return Err(StashNameError::InvalidCharacterLength(s.len()));
        }

        if s.as_bytes()[0].is_ascii_digit() {
            return Err(StashNameError::NameMustStartWithLetter);
        }

        Ok(Self(s.to_string()))
    }
}

impl ToString for StashName {
    fn to_string(&self) -> String {
        self.0.to_string()
    }
}

impl<'de> Deserialize<'de> for StashName {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        let display_name = StashName::from_str(&s).map_err(|e| serde::de::Error::custom(e.to_string()))?;
        Ok(display_name)
    }
}
