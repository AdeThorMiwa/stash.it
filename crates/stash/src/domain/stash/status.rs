use std::str::FromStr;

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum StashStatus {
    ACTIVE,
    PAUSED,
    CLOSED,
}

impl FromStr for StashStatus {
    type Err = &'static str;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "ACTIVE" => Self::ACTIVE,
            "PAUSED" => Self::PAUSED,
            "CLOSED" => Self::CLOSED,
            _ => return Err("invalid stash status"),
        })
    }
}
