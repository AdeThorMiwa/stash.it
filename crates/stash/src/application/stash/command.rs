use crate::domain::stash::{name::StashName, status::StashStatus, tag::Tag};
use shared::domain::value_objects::{mula::Mula, pid::Pid};

pub struct CreateStashCommand {
    pub user_id: Pid,
    pub name: StashName,
    pub tags: Vec<Tag>,
}

pub struct GetStashCommand {
    pub stash_id: Pid,
}

pub struct UpdateStashStatusCommand {
    pub stash_id: Pid,
    pub new_status: StashStatus,
}

pub struct UpdateStashBalanceCommand {
    pub stash_id: Pid,
    pub new_balance: Mula,
}
