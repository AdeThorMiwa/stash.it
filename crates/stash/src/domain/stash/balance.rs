use shared::domain::value_objects::{asset::Asset, mula::Mula};

#[derive(Debug, Clone)]
pub struct StashBalance {
    amount: Mula,
    asset: Asset,
}

impl StashBalance {
    pub fn new(asset: &Asset, amount: &Mula) -> Self {
        Self {
            asset: asset.clone(),
            amount: amount.clone(),
        }
    }

    pub fn get_asset(&self) -> &Asset {
        &self.asset
    }

    pub fn update_amount(&mut self, new_amount: &Mula) {
        self.amount = new_amount.clone()
    }
}
