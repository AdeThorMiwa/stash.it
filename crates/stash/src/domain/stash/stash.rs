use crate::domain::stash::{name::StashName, status::StashStatus, tag::Tag};
use chrono::Utc;
use serde_json::Value;
use shared::domain::value_objects::{date::Date, mula::Mula, pid::Pid};
use std::collections::HashMap;

pub type StashMetadata = HashMap<String, Value>;

#[derive(Debug, Clone)]
pub struct Stash {
    pid: Pid,
    user_id: Pid,
    name: StashName,
    status: StashStatus,
    tags: Vec<Tag>,
    balances: Vec<Mula>,
    metadata: StashMetadata,
    #[allow(dead_code)]
    created_at: Date,
    updated_at: Date,
}

impl Stash {
    pub fn new(user_id: &Pid, name: &StashName, tags: &Vec<Tag>) -> Self {
        Self {
            pid: Pid::new(),
            user_id: user_id.clone(),
            name: name.clone(),
            status: StashStatus::ACTIVE,
            tags: tags.clone(),
            balances: Vec::new(),
            metadata: HashMap::new(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
        }
    }

    pub fn get_pid(&self) -> &Pid {
        &self.pid
    }

    pub fn get_name(&self) -> &StashName {
        &self.name
    }

    pub fn get_user_id(&self) -> &Pid {
        &self.user_id
    }

    pub fn get_status(&self) -> &StashStatus {
        &self.status
    }

    pub fn get_tags(&self) -> &Vec<Tag> {
        &self.tags
    }

    pub fn get_balances(&self) -> &Vec<Mula> {
        &self.balances
    }

    pub fn get_metadata(&self) -> &StashMetadata {
        &self.metadata
    }

    pub fn update_name(&mut self, new_name: &StashName) {
        self.name = new_name.clone();
        self.updated_at = Utc::now();
    }

    pub fn update_status(&mut self, new_status: &StashStatus) {
        self.status = new_status.clone();
        self.updated_at = Utc::now();
    }

    pub fn update_balance(&mut self, new_balance: &Mula) {
        if let Some(balance) = self.balances.iter_mut().find(|b| b.get_asset().eq(new_balance.get_asset())) {
            *balance = new_balance.clone();
        } else {
            self.balances.push(new_balance.clone());
        }
        self.updated_at = Utc::now();
    }
}
