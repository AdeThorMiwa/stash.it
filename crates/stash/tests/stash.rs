use insta::{assert_debug_snapshot, with_settings};
use shared::{
    configure_insta,
    domain::value_objects::{asset::Asset, mula::Mula, pid::Pid},
    infrastructure::types::Result,
    testing::insta_filters::redactions::cleanup_model_generics,
};
use stash::{
    application::stash::{
        StashService,
        command::{CreateStashCommand, UpdateStashBalanceCommand, UpdateStashStatusCommand},
    },
    domain::stash::{name::StashName, status::StashStatus, tag::Tag},
};
use std::str::FromStr;

use crate::utils::{bootstrap::bootstrap, prepare::prepare_stash};

mod utils;

#[tokio::test]
async fn can_create_stash() -> Result<()> {
    // Arrange
    configure_insta!();
    let provider = bootstrap();
    let stash_service = provider.get_required::<StashService>();
    let command = CreateStashCommand {
        name: StashName::from_str("General").unwrap(),
        user_id: Pid::new(),
        tags: vec![Tag::from_str("personal").unwrap()],
    };

    // Act
    let stash = stash_service.create_stash(command).await?;

    // Assert
    with_settings!({
        filters => cleanup_model_generics(), snapshot_suffix => "create_stash"
    }, {
        assert_debug_snapshot!(stash)
    });

    Ok(())
}

#[tokio::test]
async fn can_update_stash_status() -> Result<()> {
    // Arrange
    let provider = bootstrap();
    let stash_service = provider.get_required::<StashService>();
    let stash = prepare_stash(&provider).await?;
    let command = UpdateStashStatusCommand {
        stash_id: stash.get_pid().to_owned(),
        new_status: StashStatus::CLOSED,
    };

    // Act
    let stash = stash_service.update_stash_status(command).await?;

    // Assert
    assert_eq!(stash.get_status(), &StashStatus::CLOSED, "Stash status must be `CLOSED`");

    Ok(())
}

#[tokio::test]
#[cfg(feature = "testing")]
async fn can_update_stash_balance() -> Result<()> {
    // Arrange
    let provider = bootstrap();
    let stash_service = provider.get_required::<StashService>();
    let stash = prepare_stash(&provider).await?;
    let new_balance = Mula::new(10, &Asset::usdt());
    let command = UpdateStashBalanceCommand {
        stash_id: stash.get_pid().to_owned(),
        new_balance: new_balance.clone(),
    };

    // Act
    let stash = stash_service.update_stash_balance(command).await?;
    let balance = stash.get_balances().iter().find(|b| b.get_asset() == new_balance.get_asset()).unwrap();

    // Assert
    assert_eq!(balance, &new_balance, "Balance must equal to new balance");

    Ok(())
}
