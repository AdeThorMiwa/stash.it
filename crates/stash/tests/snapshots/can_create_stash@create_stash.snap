---
source: crates/stash/tests/stash.rs
expression: stash
---
Stash {
    pid: Pid(
        PID,
    ),
    user_id: Pid(
        PID,
    ),
    name: StashName(
        "General",
    ),
    status: ACTIVE,
    tags: [
        Tag(
            "personal",
        ),
    ],
    balances: [],
    metadata: {},
    created_at: DATEZ,
    updated_at: DATEZ,
}
